components/codec/src/columnar/timestamp.rs

```diff
@@ -11,3 +11,301 @@
+                    if num == 1 {
+                        writer.write_bit(Bit(0))
+                    } else {
+                        writer.write_bit(Bit(1))
+                    }
+                    continue;
+                }
+                let cur_delta = v - last_ts;
+                let delta_of_delta = cur_delta - last_delta;
+                match TimestampEncoder::encode_delta_of_delta_masks(delta_of_delta) {
+                    DeltaOfDeltaEncodeMasks::Zero() => {
+                        writer.write_bit(Bit(0));
+                        dod_encode_ts_len += 1;
+                    }
+                    DeltaOfDeltaEncodeMasks::Normal(
+                        control_bis,
+                        control_bits_len,
+                        dod_bits_len,
+                    ) => {
+                        writer.write_bits(control_bis as u64, control_bits_len as u32);
+                        writer.write_bits(delta_of_delta as u64, dod_bits_len as u32);
+                        dod_encode_ts_len += 1;
+                    }
+                    DeltaOfDeltaEncodeMasks::None() => {
+                        writer.write_bits(END_MARKER, NUM_BITS_END_MARKER_LEN as u32);
+
+                        let fallback_ts_num = num - dod_encode_ts_len;
+                        writer.write_bits(fallback_ts_num as u64, 32);
+                        writer.write_bits(v as u64, 64);
+                        need_fallback = true;
+                    }
+                };
+                last_delta = cur_delta;
+                last_ts = v;
+
+                if num == dod_encode_ts_len {
+                    // write end mask
+                    writer.write_bits(END_MARKER, NUM_BITS_END_MARKER_LEN as u32);
+                    writer.write_bits(0, 32);
+                    continue;
+                }
+            } else {
+                // Unable to encode with dod, break out of loop and fallback to i64 storage
+                writer.write_bits(v as u64, 64);
+            }
+        }
+
+        buf.put_slice(writer.close().as_ref());
+        Ok(())
+    }
+}
+
+impl TimestampDecoder {
+    fn encode_delta_of_delta_masks(dod_control_bits_size: i32) -> DeltaOfDeltaEncodeMasks {
+        match dod_control_bits_size {
+            0 => DeltaOfDeltaEncodeMasks::Zero(),
+            1 => DeltaOfDeltaEncodeMasks::Normal(0b10, 2, 7),
+            2 => DeltaOfDeltaEncodeMasks::Normal(0b110, 3, 9),
+            3 => DeltaOfDeltaEncodeMasks::Normal(0b1110, 4, 12),
+            4 => DeltaOfDeltaEncodeMasks::Normal(0b1111, 4, 32),
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn decode<B, F>(buf: &B, mut f: F) -> Result<()>
+    where
+        B: Buf,
+        F: FnMut(Timestamp) -> Result<()>,
+    {
+        let mut reader = BufferedReader::new(buf.chunk());
+
+        let version = reader
+            .next_bits((NUM_BYTES_ENCODE_VERSION_LEN * 8) as u32)
+            .context(ReadEncode)? as u8;
```

Comment:
The `version` can be retrieved by `buf.get_u8()`.

Reply:
Here, the reader is used uniformly to read data, and in order to read subsequent data, we need to advance the inner cursor inside the reader.

Reply:
After `buf.get_u8()`, the remaining bytes can be used to build the bit reader.